File: content/snippets/python/design_patterns_adapter.py
from dataclasses import dataclass


class Printer:
    def print(self, message: str) -> None:
        raise NotImplementedError


class JsonLogger:
    def write_json(self, payload: str) -> None:
        print(payload)


@dataclass
class JsonLoggerAdapter(Printer):
    logger: JsonLogger

    def print(self, message: str) -> None:
        self.logger.write_json(f'{{"message":"{message}"}}')


JsonLoggerAdapter(JsonLogger()).print("build finished")

File: content/snippets/typescript/design_patterns_adapter.ts
interface Printer {
  print(message: string): void;
}

class JsonLogger {
  writeJson(payload: string): void {
    console.log(payload);
  }
}

class JsonLoggerAdapter implements Printer {
  constructor(private readonly logger: JsonLogger) {}

  print(message: string): void {
    this.logger.writeJson(`{"message":"${message}"}`);
  }
}

new JsonLoggerAdapter(new JsonLogger()).print('build finished');
